src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+  //
+  for (unsigned int ix = 0; ix < condVec.size(); ++ix)
+  {
+    BSONObj                   condition = condVec[ix].embeddedObject();
+    std::string               condType;
+    std::vector<BSONElement>  valueVec;
+
+    condType = condition.getStringField(CSUB_CONDITIONS_TYPE);
+    if (condType != "ONCHANGE")
+    {
+      continue;
+    }
+
+    NotifyCondition* ncP = new NotifyCondition();
+    ncP->type = condType;
+
+    valueVec = condition.getField(CSUB_CONDITIONS_VALUE).Array();
+    for (unsigned int vIx = 0; vIx < valueVec.size(); ++vIx)
+    {
+      std::string condValue;
+
+      condValue = valueVec[vIx].String();
+      ncP->condValueList.push_back(condValue);
+    }
+
+    cSubP->notifyConditionVector.push_back(ncP);
+  }
+
+  if (cSubP->notifyConditionVector.size() == 0)  // Cleanup
+  {
+    LM_E(("ERROR (empty notifyConditionVector) - cleaning up"));
+    cachedSubscriptionDestroy(cSubP);
+    delete cSubP;
+    return -4;
+  }
+
+  mongoSubCacheItemInsert(cSubP);
+
+  return 0;
+}
+
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemInsert - create a new sub, fill it in, and add it to cache
+*/
+void mongoSubCacheItemInsert
+(
+  const char*               tenant,
+  const char*               servicePath,
+  SubscribeContextRequest*  scrP,
+  const char*               subscriptionId,
+  int64_t                   expirationTime,
+  int64_t                   throttling,
+  Format                    notifyFormat
+)
+{
+  CachedSubscription* cSubP = new CachedSubscription();
+  LM_T(LmtMongoSubCache,  ("allocated CachedSubscription at %p", cSubP));
+
+  //
+  // 1. First the non-complex values
+  //
+  cSubP->tenant                = (tenant[0] == 0)? NULL : strdup(tenant);
+  cSubP->servicePath           = strdup(servicePath);
+  cSubP->subscriptionId        = strdup(subscriptionId);
+  cSubP->reference             = strdup(scrP->reference.get().c_str());
+  cSubP->expirationTime        = expirationTime;
+  cSubP->throttling            = throttling;
+  cSubP->lastNotificationTime  = 0;
+  cSubP->pendingNotifications  = 0;
+  cSubP->notifyFormat          = notifyFormat;
+  cSubP->next                  = NULL;
+
+  LM_T(LmtMongoSubCache, ("inserting a new sub in cache (%s). lastNotifictionTime: %lu", cSubP->subscriptionId, cSubP->lastNotificationTime));
+
+
+  //
+  // 2. Then the values that have functions/methods for filling/parsing
+  //
+  cSubP->throttling = scrP->throttling.parse();
+  cSubP->notifyConditionVector.fill(scrP->notifyConditionVector);
+
+
+  //
+  // 3. Convert all patterned EntityId to EntityInfo
+  //
+  for (unsigned int ix = 0; ix < scrP->entityIdVector.vec.size(); ++ix)
+  {
+    EntityId* eIdP;
+
+    eIdP = scrP->entityIdVector.vec[ix];
+
+    if (eIdP->isPattern == "false")
+    {
+      continue;
+    }
+
+    EntityInfo* eP = new EntityInfo(eIdP->id, eIdP->type);
+    
+    cSubP->entityIdInfos.push_back(eP);
+  }
+
```

Comment:
What about if the entityIdVector doesn't contain any pattern subscription? Maybe the entityIdInfo vector size should be checked (if 0, then exist without effective insertion).